module/builder/collection/builder.go

```diff
@@ -119,41 +147,219 @@ func (b *Builder) BuildOn(parentID flow.Identifier, setter func(*flow.Header) er
+	span, _ = b.tracer.StartSpanFromContext(ctx, trace.COLBuildOnUnfinalizedLookup)
+	err = b.populateUnfinalizedAncestryLookup(buildCtx)
+	span.End()
 	if err != nil {
 		return nil, fmt.Errorf("could not populate un-finalized ancestry lookout (parent_id=%x): %w", parentID, err)
 	}
 
-	// TODO (ramtin): enable this again
-	// b.tracer.FinishSpan(parentID, trace.COLBuildOnUnfinalizedLookup)
-	// b.tracer.StartSpan(parentID, trace.COLBuildOnFinalizedLookup)
-	// defer b.tracer.FinishSpan(parentID, trace.COLBuildOnFinalizedLookup)
-
 	// STEP 1b: create a lookup of all transactions previously included in
 	// the finalized collections. Any transactions already included in finalized
 	// collections can be removed from the mempool.
-	err = b.populateFinalizedAncestryLookup(buildCtx.lowestPossibleReferenceBlockHeight(), buildCtx.highestPossibleReferenceBlockHeight(), lookup, limiter)
+	span, _ = b.tracer.StartSpanFromContext(ctx, trace.COLBuildOnFinalizedLookup)
+	err = b.populateFinalizedAncestryLookup(buildCtx)
+	span.End()
 	if err != nil {
 		return nil, fmt.Errorf("could not populate finalized ancestry lookup: %w", err)
 	}
 
-	// TODO (ramtin): enable this again
-	// b.tracer.FinishSpan(parentID, trace.COLBuildOnFinalizedLookup)
-	// b.tracer.StartSpan(parentID, trace.COLBuildOnCreatePayload)
-	// defer b.tracer.FinishSpan(parentID, trace.COLBuildOnCreatePayload)
-
-	// STEP TWO: build a payload of valid transactions, while at the same
+	// STEP 2: build a payload of valid transactions, while at the same
 	// time figuring out the correct reference block ID for the collection.
+	span, _ = b.tracer.StartSpanFromContext(ctx, trace.COLBuildOnCreatePayload)
+	payload, err := b.buildPayload(buildCtx)
+	span.End()
+	if err != nil {
+		return nil, fmt.Errorf("could not build payload: %w", err)
+	}
+
+	// STEP 3: we have a set of transactions that are valid to include on this fork.
+	// Now we create the header for the cluster block.
+	span, _ = b.tracer.StartSpanFromContext(ctx, trace.COLBuildOnCreateHeader)
+	header, err := b.buildHeader(buildCtx, payload, setter)
+	span.End()
+	if err != nil {
+		return nil, fmt.Errorf("could not build header: %w", err)
+	}
+
+	proposal := cluster.Block{
+		Header:  header,
+		Payload: payload,
+	}
+
+	// STEP 4: insert the cluster block to the database.
+	span, _ = b.tracer.StartSpanFromContext(ctx, trace.COLBuildOnDBInsert)
+	err = operation.RetryOnConflict(b.db.Update, procedure.InsertClusterBlock(&proposal))
+	span.End()
+	if err != nil {
+		return nil, fmt.Errorf("could not insert built block: %w", err)
+	}
+
+	return proposal.Header, nil
+}
+
+// getBlockBuildContext retrieves the required contextual information from the database
+// required to build a new block proposal.
+// No errors are expected during normal operation.
+func (b *Builder) getBlockBuildContext(parentID flow.Identifier) (*blockBuildContext, error) {
+	ctx := new(blockBuildContext)
```

Comment:
I haven't seen this frequently, any specific reason for doing this instead of:
```golang
:= &blockBuildContext {
...
}
```

Reply:
The (very minor) benefit of this over struct initialization is it makes it more clear that the variable declaration isn't fully constructing the object.

Relatively obvious that `ctx` is not completely constructed yet, because it is zeroed.
```go
ctx := new(blockBuildContext)
```

Not obvious that `ctx` is not completely constructed yet, unless you know how many fields `ctx` has:
```go
ctx := &blockBuildContext{
  config: b.config,
  parentID: parentID,
  lookup: newLookup(),
}
```